packages/opentelemetry-core/src/trace/instrumentation/BasePlugin.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Tracer, Plugin } from '@opentelemetry/types';
+
+/** This class represent the base to patch plugin. */
+export abstract class BasePlugin implements Plugin {
+  // tslint:disable-next-line:no-any
+  protected moduleExports: any;
+  protected tracer!: Tracer;
+
+  enable<T>(moduleExports: T, tracer: Tracer): T {
+    this.moduleExports = moduleExports;
+    this.tracer = tracer;
+    return this.applyPatch();
+  }
+
+  disable(): void {
+    this.applyUnpatch();
+  }
+
+  // tslint:disable-next-line:no-any
+  protected abstract applyPatch(): any;
```

Comment:
Why does `applyPatch` need to return anything? Could it just be `void`?

Reply:
The use case I can think of is that it allows to override the export if it's a function since the only way to patch a function is to completely replace the export. Then the loader could replace the export by whatever is returned by `patch`.